File: 1512-design-underground-system/design-underground-system.py
class UndergroundSystem:

    def __init__(self):
        self.checkins = {}
        self.averages = {}
        

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.checkins[id] = self.Checkin(id,stationName,t)
        
    def checkOut(self, id: int, stationName: str, t: int) -> None:
        timeDif = t - self.checkins[id].t
        stations = self.checkins[id].stationName + "," + stationName

        average = self.averages.get(stations,self.Average())

        average.updateAverage(timeDif)

        self.averages[stations] = average
        
        self.checkins.pop(id)
        

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        stations = startStation + "," + endStation
        return self.averages[stations].getAverage()

    

    class Checkin:

        def __init__(self, id: int, stationName: str, t: int):
            self.id = id
            self.stationName = stationName
            self.t = t

    class Average:
        def __init__(self):
            self.totalTime = 0
            self.customers = 0

        def updateAverage(self,timeDif):
            self.totalTime+=timeDif
            self.customers+=1
        
        def getAverage(self):
            return self.totalTime/self.customers
        

# SC -> O(N+M)
# TC -> O(1)

# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

File: 1512-design-underground-system/design-underground-system.java
class UndergroundSystem {
    Map<Integer,Journey> journeys; 
    Map<String,Average> averages;

    public UndergroundSystem() {
        journeys = new HashMap<>();
        averages = new HashMap<>();
    }
    
    public void checkIn(int id, String stationName, int t) {
        this.journeys.put(id,new Journey(id,stationName,t));
    }
    
    public void checkOut(int id, String stationName, int t) {
        Journey journey = this.journeys.get(id);
        String stations = journey.stationName + "," + stationName;
        int timeDif = t - journey.t;

        Average average = this.averages.containsKey(stations) ? this.averages.get(stations) : new Average();

        average.updateAverage(timeDif);

        this.averages.put(stations,average);

        this.journeys.remove(id);

    }
    
    public double getAverageTime(String startStation, String endStation) {
        String stations = startStation + "," + endStation;
        return this.averages.get(stations).getAverage();
        
    }

    class Journey { 
        int id;
        String stationName;
        int t;

        public Journey(int id, String stationName,int t){
            this.id = id;
            this.stationName = stationName;
            this.t = t;
        }


    }

    class Average {
        double totalTime = 0;
        int customers = 0;

        public void updateAverage(int timeDif){
            totalTime+=timeDif;
            customers+=1;
        }

        public double getAverage(){
            return totalTime / customers;
        }

    }

}

// SC -> O(N+M)
// TC -> O(1)

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */